src/lib.rs

```diff
@@ -69,75 +110,87 @@ pub unsafe extern "C" fn pthread_join(
     native: libc::pthread_t,
     _value: *mut *mut libc::c_void,
 ) -> libc::c_int {
-    if native == MAIN_THREAD_ID {
-        // This is not a valid thread to join on
-        return libc::EINVAL;
-    }
+    let pthread = Box::from_raw(native as *mut PThread);
 
-    let result = ctru_sys::threadJoin(native as ctru_sys::Thread, u64::MAX);
+    let result = ctru_sys::threadJoin(pthread.thread, u64::MAX);
     if ctru_sys::R_FAILED(result) {
         // TODO: improve the error code by checking the result further?
         return libc::EINVAL;
     }
 
-    ctru_sys::threadFree(native as ctru_sys::Thread);
+    ctru_sys::threadFree(pthread.thread);
 
     0
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn pthread_detach(thread: libc::pthread_t) -> libc::c_int {
-    if thread == MAIN_THREAD_ID {
-        // This is not a valid thread to detach
-        return libc::EINVAL;
-    }
-
-    ctru_sys::threadDetach(thread as ctru_sys::Thread);
+pub unsafe extern "C" fn pthread_detach(native: libc::pthread_t) -> libc::c_int {
+    let pthread = Box::from_raw(native as *mut PThread);
+    ctru_sys::threadDetach(pthread.thread);
 
     0
 }
 
 #[no_mangle]
 pub unsafe extern "C" fn pthread_self() -> libc::pthread_t {
-    ctru_sys::threadGetCurrent() as libc::pthread_t
-}
-
-unsafe fn get_main_thread_handle() -> Result<ctru_sys::Handle, ctru_sys::Result> {
-    // Unfortunately I don't know of any better way to get the main thread's
-    // handle, other than via svcGetThreadList and svcOpenThread.
-    // Experimentally, the main thread ID is always the first in the list.
-    let mut thread_ids = [0; 1];
-    let mut thread_ids_count = 0;
-    let result = ctru_sys::svcGetThreadList(
-        &mut thread_ids_count,
-        thread_ids.as_mut_ptr(),
-        thread_ids.len() as i32,
-        ctru_sys::CUR_PROCESS_HANDLE,
-    );
-    if ctru_sys::R_FAILED(result) {
-        return Err(result);
+    if let Some(thread) = THREAD_SELF {
+        return thread as libc::pthread_t;
     }
 
-    // Get the main thread's handle
-    let main_thread_id = thread_ids[0];
-    let mut main_thread_handle = 0;
-    let result = ctru_sys::svcOpenThread(
-        &mut main_thread_handle,
-        ctru_sys::CUR_PROCESS_HANDLE,
-        main_thread_id,
-    );
-    if ctru_sys::R_FAILED(result) {
-        return Err(result);
+    let thread = ctru_sys::threadGetCurrent();
+    let mut os_thread = 0;
+    let res = ctru_sys::svcGetThreadId(&mut os_thread, ctru_sys::CUR_THREAD_HANDLE);
+
+    if ctru_sys::R_FAILED(res) {
+        // TODO: some way to report error? This should probably normally never fail...
+        ptr::null_mut::<PThread>() as libc::pthread_t
+    } else {
+        let pthread = Box::new(PThread { thread, os_thread });
+        let raw = Box::into_raw(pthread);
+        THREAD_SELF = Some(raw);
+        raw as libc::pthread_t
```

Comment:
After looking at the [pthread_self](https://man7.org/linux/man-pages/man3/pthread_self.3.html) documentation, I think this approach is flawed. We can't return a new allocation here. It has to be the same ID that we returned in `pthread_create`:
> The pthread_self() function returns the ID of the calling thread.
       This is the same value that is returned in *thread in the
       [pthread_create(3)](https://man7.org/linux/man-pages/man3/pthread_create.3.html) call that created this thread.

Reply:
We may be able to work around this by setting the value of `THREAD_SELF` before calling `entrypoint` in `pthread_create`. This is how libctru works.

Reply:
I thought about that and it would be nicer, but I'm not sure if there's an easy way to set the value of a `#[thread_local]` for a _different_ thread. Would the value be copied if it was set before actually starting `entrypoint`?

Reply:
Yeah, I mean to say do this in the new thread before calling entrypoint.

Reply:
See the comments in `pthread_detach`, I don't think using thread locals as the main storage is going to work.